snark-verifier-sdk/src/halo2/aggregation.rs

```diff
@@ -0,0 +1,458 @@
+)
+where
+    AS: PolynomialCommitmentScheme<
+            G1Affine,
+            Rc<Halo2Loader<'a>>,
+            VerifyingKey = Svk,
+            Output = KzgAccumulator<G1Affine, Rc<Halo2Loader<'a>>>,
+        > + AccumulationScheme<
+            G1Affine,
+            Rc<Halo2Loader<'a>>,
+            Accumulator = KzgAccumulator<G1Affine, Rc<Halo2Loader<'a>>>,
+            VerifyingKey = KzgAsVerifyingKey,
+        >,
+{
+    assert!(!snarks.is_empty(), "trying to aggregate 0 snarks");
+    let assign_instances = |instances: &[Vec<Value<Fr>>]| {
+        instances
+            .iter()
+            .map(|instances| {
+                instances
+                    .iter()
+                    .map(|instance| loader.assign_scalar(*instance))
+                    .collect_vec()
+            })
+            .collect_vec()
+    };
+
+    let mut previous_instances = Vec::with_capacity(snarks.len());
+    let mut accumulators = snarks
+        .iter()
+        .flat_map(|snark| {
+            let protocol = snark.protocol.loaded(loader);
+            let instances = assign_instances(&snark.instances);
+            // read the transcript and perform Fiat-Shamir
+            // run through verification computation and produce the final pair `succinct`
+            let mut transcript = PoseidonTranscript::<Rc<Halo2Loader>, _>::from_spec(
+                loader,
+                snark.proof(),
+                POSEIDON_SPEC.clone(),
+            );
+            let proof = PlonkSuccinctVerifier::<AS>::read_proof(
+                svk,
+                &protocol,
+                &instances,
+                &mut transcript,
+            )
+            .unwrap();
+            let accumulator =
+                PlonkSuccinctVerifier::<AS>::verify(svk, &protocol, &instances, &proof).unwrap();
+
+            previous_instances.push(
+                instances
+                    .into_iter()
+                    .flatten()
+                    .map(|scalar| scalar.into_assigned())
+                    .collect(),
+            );
+
+            accumulator
+        })
+        .collect_vec();
+
+    let accumulator = if accumulators.len() > 1 {
+        let mut transcript = PoseidonTranscript::<Rc<Halo2Loader>, _>::from_spec(
+            loader,
+            as_proof,
+            POSEIDON_SPEC.clone(),
+        );
+        let proof = <AS as AccumulationScheme<_, _>>::read_proof(
+            &Default::default(),
+            &accumulators,
+            &mut transcript,
+        )
+        .unwrap();
+        <AS as AccumulationScheme<_, _>>::verify(&Default::default(), &accumulators, &proof)
+            .unwrap()
+    } else {
+        accumulators.pop().unwrap()
+    };
+
+    (previous_instances, accumulator)
+}
+
+/// Same as `FlexGateConfigParams` except we assume a single Phase and default 'Vertical' strategy.
+/// Also adds `lookup_bits` field.
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub struct AggregationConfigParams {
```

Comment:
Seems not used yet, perhaps we could remove this?

Reply:
Ah yes, left over from our version